consensus/src/chained_bft/liveness/multi_proposer_election.rs

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::chained_bft::{
+    common::{Author, Payload, Round},
+    consensus_types::block::Block,
+    liveness::proposer_election::ProposerElection,
+};
+
+use siphasher::sip::SipHasher13;
+use std::hash::{Hash, Hasher};
+
+/// The MultiProposer maps a round to an ordered list of authors.
+/// The primary proposer is determined by an index of hash(round) % num_proposers.
+/// The secondary proposer is determined by hash(hash(round)) % (num_proposers - 1), etc.
+/// In order to ensure the required number of proposers, a set of the proposers to choose from
+/// is updated after each hash: a chosen candidate is removed to avoid duplication.
+///
+/// Note the hash doesn't have to be cryptographic. The goal is to make sure that different
+/// combinations of consecutive leaders are going to appear with equal probability.
+
+/// While each round has more than a single valid proposer, only the primary proposer is
+/// considered for `process_proposal`.
+/// TODO: we're going to extend this functionality with an ability to use secondary proposals upon
+/// timeout.
+pub struct MultiProposer {
+    // Ordering of proposers to rotate through (all honest replicas must agree on this)
+    proposers: Vec<Author>,
+    // Number of proposers per round
+    num_proposers_per_round: usize,
+}
+
+impl MultiProposer {
+    pub fn new(proposers: Vec<Author>, num_proposers_per_round: usize) -> Self {
+        assert!(num_proposers_per_round > 0);
+        assert!(num_proposers_per_round <= proposers.len());
+        Self {
+            proposers,
+            num_proposers_per_round,
+        }
+    }
+
+    // A deterministic hashing function based on SipHash 1-3 hasher
+    pub fn hash(val: u64) -> u64 {
+        let mut hasher = SipHasher13::new();
```

Comment:
I have no idea about SipHasher, do we need to create one instance per call?

Reply:
Yea, apparently this is how any hash function works today in Rust.